Improved Lit handling. Functional extensionality tests

amaurremi opened the pull request:
This PR is part of the work on improving lambda handling in Dafny. A [previous](https://github.com/boogie-org/boogie/commit/99c51f6b0a6aa4928f4cabb724e290ff0cb82b97) Boogie commit generalizes Boogie's lambda lifting algorithm which allows for a more uniform representation of anonymous functions, allowing one to (more likely) have the same representation of lambdas that are created in different contexts. 

This PR ensures that `const`s do not get wrapped in `Lit`s (for example, because `const`s can be uninitialized), which also helps to make lambda lifting more uniform: if two syntactically equal lambdas have a free variable `x`, where in one case `x` is a `const` and in the other a parameter, then at the moment, the first lambda will contain a `Lit` whereas the latter one will not. After this PR, both lambdas would not have a `Lit` which will allow them to be lifted in the same way.

Since lambda lifting makes lambdas more usable, it also allows us to use functional extensionality (if we add it as an axiom). This PR adds two tests with proofs that make use of functional extensionality: a proof of associativity of matrix multiplication where matrices are represented as functions, and monad laws for the `State` monad.

@gancherj @RustanLeino @samuelgruetter @seanmcl

RustanLeino commented on Test/lambdas/LitInt.dfy:
```diff
@@ -0,0 +1,22 @@
+
+function Fibonacci(n: int): int
+{
+	if n <= 1 then 1
```
This is not the Fibonacci I know, because it skips 0. I suggest:

    if n < 2 then n

samuelgruetter commented on Source/Dafny/Translator.cs:
```diff
@@ -14381,12 +14380,8 @@ public Expression GetSubstitutedBody(LetExpr e)
                 result = translator.CondApplyUnbox(expr.tok, result, field.Type, expr.Type);
-                if (isLit) {
-                  result = MaybeLit(result, translator.TrType(expr.Type));
-                }
```
If I understand correctly, this change results in fewer `Lit`s. Does that mean there will be cases where the SMT solver could compute thanks to `Lit`s before but now cannot any more?

amaurremi replied:
Yes, you are right, that's a good point. I added a test case to illustrate the difference in expressivity.
The reason that Rustan pointed out to me why we still want to not have `Lit`s around `const`s (and not, for example, check whether the right-hand side of a `const` assignment has a `Lit`) is that suppose we have
```ml
module Const {
  export provides N
  const N := 3;
}

module Test {
import C = Const

  lemma test() {
    assert C.N == 3; // will fail
  }
}
```
The assertion will fail because the definition of `N` is hidden inside the `Const` module. If `N` were wrapped inside a `Lit` then in the outside module we might try to evaluate expressions that use `N` which we want to prevent. Therefore, we don't want `const`s to be `Lit`-wrapped at all.

RustanLeino commented on Test/lambdas/LitInt.dfy:
```diff
@@ -0,0 +1,22 @@
+function Fibonacci(n: int): int
+{
+	if n <= 1 then 1
+	else Fibonacci(n - 2) + Fibonacci(n - 1)
```
The indentation looks suspiciously big. You're not using tabs, right? I suggest 2 spaces for the indent.

RustanLeino commented on Test/lambdas/LitInt.dfy:
```diff
@@ -0,0 +1,22 @@
+// since that would not terminate. In order to avoid that
+// we do not wrap `const`s in `Lit`s. If we did wrap this
+// `N` in a `Lit` the verifier would keep trying to prove this
+// assertion.
```
Nice comment

amaurremi replied:
Thanks!

RustanLeino commented on Test/lambdas/MatrixAssoc.dfy:
```diff
@@ -0,0 +1,270 @@
+    FunEq(a, b);
+}
+
+/** Σ from i to n of f(i) */
```
The comment should say "`0` to `n`" (which leaves `i` rather unbound in the sentence). Perhaps

    Σ f(i) where i ranges from 0 to n

On the subject of "`0` to `n`":  One way to distinguish half-open intervals (like `[0..100]` in Dafny, `[0, 100)` in textbooks, or `i: 0 <= i < 100` using a predicate) from closed intervals (like `[0, 99]` in textbooks or `i: 0 <= i <= 99`) is to use "`0` to `100`" for the former and "`0` through `99`" for the latter. The mnemonic for these is that "to" tells you to stop as soon as you reach `100`, whereas "through" says you go all the way through `99` before you stop. Using this convention (which I'm not claiming is widespread, but it makes sense to me), "`0` to `n`" is indeed what you want. But if you want to emphasize that more, you could say (the grammatically dubious, but unambiguous) "`0` to less than `n`".

RustanLeino commented on Test/lambdas/MatrixAssoc.dfy:
```diff
@@ -0,0 +1,270 @@
+    ensures Sum_n((k: Index) => Sum_n((l: Index) => m(k)(l), n1), n2)
+            ==
+            Sum_n((l: Index) => Sum_n((k: Index) => m(k)(l), n2), n1)
+    {
```
Bring this open-curly to the left margin, lining up with `lemma`.

RustanLeino commented on Test/lambdas/MatrixAssoc.dfy:
```diff
@@ -0,0 +1,270 @@
+    {
+        var f  := (k: Index) => Sum_n((l: Index) => m(k)(l), n1);
+        var g  := (l: Index) => Sum_n((k: Index) => m(k)(l), n2);
+        if (n2 == 0) {
```
Parens not needed

RustanLeino commented on Test/lambdas/MatrixAssoc.dfy:
```diff
@@ -0,0 +1,270 @@
+        if (n2 == 0) {
+            calc {
+                Sum_n(f, n2);
+                ==
```
I suggest lining up the `==` with `calc`. (Something should be said in the Dafny style guide.)

amaurremi replied:
Those two test files had pretty careless formatting, I tried to reformat it into something clearer and more consistent now.

RustanLeino commented on Test/lambdas/StateMonad.dfy:
```diff
@@ -0,0 +1,231 @@
+    }
+
+  lemma Associativity<A,B,C>(m : M<A>, f:A -> M<B>, g: B -> M<C>)
+  ensures Bind(Bind(m, f), g) == Bind(m, x => Bind(f(x), g))
```
Indent the `ensures`

seanmcl commented on Test/lambdas/StateMonad.dfy:
```diff
@@ -0,0 +1,231 @@
+  { assume false; }
+
+  lemma LeftIdentity<A,B>(x : A, f : A -> M<B>)
+    ensures Bind(Return(x), f) == f(x) // return x >>= f  ==  f x
```
1 space around == in comment

amaurremi replied:
I removed that comment now since it's repeating another comment from above.

seanmcl commented on Test/lambdas/StateMonad.dfy:
```diff
@@ -0,0 +1,231 @@
+        }
+    }
+
+  lemma Associativity<A,B,C>(m : M<A>, f:A -> M<B>, g: B -> M<C>)
```
Not a criticism, but curious: is there a path to making this kind of proof less verbose?  I'm not used to seeing so many manual expansions in calc sequences.  Do you think this is a fundamental limitation, or do we just need more axioms or triggers or something in the way Dafny automates other features.

amaurremi replied:
I think that part of what makes the proof so verbose is that we are reasoning about nested lambdas each of which we have to write out in full because the verifier doesn't reduce anything inside of lambda bodies on its own. So not only do we have to keep writing everything fully, but we have to also copy-paste the lambdas into the functional-extensionality-axiom application `FunEq`.
According to @RustanLeino even when I add `FunEq` to Dafny by default (so that we can write `f == g` rather than `FunEq(f, g)`) we will still have to explicitly give hints to the verifier that it should invoke functional extensionality, which will not make the proofs much less verbose. Although I do hope that it'll be able to apply the axiom on its own, at least sometimes, but I am still in the process of adding extensionality by default so I don't know yet.

amaurremi commented:
@RustanLeino @samuelgruetter @seanmcl Thanks a lot for all your comments, I passed through them in two [additional](https://github.com/dafny-lang/dafny/pull/344/commits/a9485a3268cc33f9497faf92a7075b5503a1833d) [commits](https://github.com/dafny-lang/dafny/pull/344/commits/a0bdda356a511aaca76a538a01aff30be8b26ec5).

samuelgruetter commented on Test/lambdas/LitInt.dfy:
```diff
@@ -0,0 +1,36 @@
+  // so Fibonacci(M+5) cannot be evaluated without a fuel
+  // (but as shown belowi in `CheckFib` the verifier knows that
+  // Fibonacci(15) == 610)
+  assert Fibonacci(M + 5) == 610;
```
`M + 5` equals 17, so this assertion does not hold according to math, so you're not testing what you claim to test? Also, typo in "belowi"

amaurremi replied:
Thanks for noticing!

amaurremi commented:
Fixed the arithmetic problem that Sam pointed out, reverted DafnyRuntime.cs, and hopefully got the documentation/spacing to an acceptable state, but I'm happy to get more suggestions.
